File: Python/day16.py
#!/usr/bin/env python3
import numpy

from util import *


def to_bits(line):
    return ''.join([format(int(c, 16), "04b") for c in line])


def parse_literal(msg, offset):
    value = ''
    while msg[offset] == '1':
        value += msg[offset + 1:offset + 5]
        offset += 5
    value += msg[offset + 1:offset + 5]
    return offset + 5, int(value, 2)


def parse_sub_by_bits(msg, offset, length):
    packets = []
    while offset < length:
        offset, value = parse_packet(msg, offset)
        packets.append(value)
    return offset, packets


def parse_sub_by_numb(msg, offset, length):
    packets = []
    while len(packets) < length:
        offset, value = parse_packet(msg, offset)
        packets.append(value)
    return offset, packets


def parse_packet(msg, offset):
    version = int(msg[offset:offset + 3], 2)
    ptype = int(msg[offset + 3:offset + 6], 2)
    offset += 6
    if ptype == 4:
        offset, value = parse_literal(msg, offset)
    elif msg[offset] == '0':
        length = int(msg[offset + 1:offset + 16], 2)
        offset += 16
        offset, value = parse_sub_by_bits(msg, offset, offset + length)
    else:
        length = int(msg[offset + 1:offset + 12], 2)
        offset += 12
        offset, value = parse_sub_by_numb(msg, offset, length)
    return offset, (version, ptype, value)


def parse(msg):
    offset, msg = parse_packet(msg, 0)
    return msg


def sum_versions(messages):
    total = 0
    for msg in messages:
        total += msg[0]
        if msg[1] != 4:
            total += sum_versions(msg[2])
    return total


def calculate_1(msg):
    ptype = msg[1]
    sub = msg[2]
    if ptype == 4:
        return sub
    subval = calculate(sub)
    if ptype == 0:
        value = numpy.sum(subval)
    elif ptype == 1:
        value = numpy.prod(subval)
    elif ptype == 2:
        value = min(subval)
    elif ptype == 3:
        value = max(subval)
    elif ptype == 5:
        value = 1 if subval[0] > subval[1] else 0
    elif ptype == 6:
        value = 1 if subval[0] < subval[1] else 0
    elif ptype == 7:
        value = 1 if subval[0] == subval[1] else 0
    else:
        raise "oops"
    return value


def calculate(messages):
    return [calculate_1(m) for m in messages]


if __name__ == "__main__":
    data = load('day16.txt')[0]
    bits = to_bits(data)
    message = parse(bits)

    version_sum = sum_versions([message])
    print('Version sum: %d' % version_sum)
    assert version_sum == 879

    result = calculate([message])
    print(f'Result: {result}')
    assert result[0] == 539051801941

File: Elixir/lib/day16.ex
defmodule Day16 do
  @moduledoc """
  Documentation for `Day16`.
  """

  @hex %{
    "0" => [0, 0, 0, 0],
    "1" => [0, 0, 0, 1],
    "2" => [0, 0, 1, 0],
    "3" => [0, 0, 1, 1],
    "4" => [0, 1, 0, 0],
    "5" => [0, 1, 0, 1],
    "6" => [0, 1, 1, 0],
    "7" => [0, 1, 1, 1],
    "8" => [1, 0, 0, 0],
    "9" => [1, 0, 0, 1],
    "A" => [1, 0, 1, 0],
    "B" => [1, 0, 1, 1],
    "C" => [1, 1, 0, 0],
    "D" => [1, 1, 0, 1],
    "E" => [1, 1, 1, 0],
    "F" => [1, 1, 1, 1],
  }

  def eat(s, 0, value), do: {value, s}
  def eat([head | tail], num, value) do
    eat(tail, num - 1, value * 2 + head)
  end
  def eat(s, num) do
    eat(s, num, 0)
  end

  def parse_literal([0 | s], value), do: eat(s, 4, value)
  def parse_literal([1 | s], value) do
    {value, s} = eat(s, 4, value)
    parse_literal(s, value)
  end

  def parse_by_bits([]), do: []
  def parse_by_bits(s) do
    {pkt, s} = parse(s)
    [pkt | parse_by_bits(s)]
  end

  def parse(s, 0) do
    {bits, s} = eat(s, 15)
    {sub, s} = Enum.split(s, bits)
    {parse_by_bits(sub), s}
  end

  def parse(s, 1) do
    {packets, s} = eat(s, 11)
    1..packets
    |> Enum.reduce(
         {[], s},
         fn _, {acc, s} ->
           {pkt, s} = parse(s)
           {acc ++ [pkt], s}
         end
       )
  end

  def parse([]), do: {nil, []}
  def parse(s) do
    {version, s} = eat(s, 3)
    {type, s} = eat(s, 3)
    {body, s} = case type do
      4 -> parse_literal(s, 0)
      _ -> {length_type, s} = eat(s, 1)
           parse(s, length_type)
    end
    {{version, type, body}, s}
  end

  def to_pkt(s) do
    {pkt, _s} = parse(s)
    pkt
  end

  def input() do
    File.read!("data/day16.txt")
    |> String.split("\n")
    |> Enum.at(0)
    |> String.split("", trim: true)
    |> Enum.flat_map(fn d -> @hex[d] end)
    |> to_pkt()
  end

  def version({v, 4, _sp}), do: v
  def version({v, _t, sp}) do
    v + Enum.sum(Enum.map(sp, &version/1))
  end

  def version() do
    version(input())
  end

  def evaluate({_v, 0, sp}), do: Enum.sum(Enum.map(sp, &evaluate/1))
  def evaluate({_v, 1, sp}), do: Enum.product(Enum.map(sp, &evaluate/1))
  def evaluate({_v, 2, sp}), do: Enum.min(Enum.map(sp, &evaluate/1))
  def evaluate({_v, 3, sp}), do: Enum.max(Enum.map(sp, &evaluate/1))
  def evaluate({_v, 4, value}), do: value
  def evaluate({_v, 5, [a | [b | _]]}) do
    v1 = evaluate(a)
    v2 = evaluate(b)
    if v1 > v2, do: 1, else: 0
  end
  def evaluate({_v, 6, [a | [b | _]]}) do
    v1 = evaluate(a)
    v2 = evaluate(b)
    if v1 < v2, do: 1, else: 0
  end
  def evaluate({_v, 7, [a | [b | _]]}) do
    v1 = evaluate(a)
    v2 = evaluate(b)
    if v1 == v2, do: 1, else: 0
  end

  def evaluate() do
    evaluate(input())
  end
end
